src/serial/private/utils/windows_registry.nim

```diff
@@ -0,0 +1,131 @@
+#
+# mp035: modified version of the registry module from the Nim standard library
+#
+
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2016 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module is experimental and its interface may change.
+
+import winlean, os
+
+type
+  HKEY* = uint
+
+const
+  HKEY_LOCAL_MACHINE* = HKEY(0x80000002u)
+  HKEY_CURRENT_USER* = HKEY(2147483649)
+
+  RRF_RT_ANY = 0x0000ffff
+  RRF_RT_REG_DWORD = 0x00000010
+  KEY_WOW64_64KEY = 0x0100
+  KEY_WOW64_32KEY = 0x0200
+  KEY_READ = 0x00020019
+  REG_SZ = 1
+
+proc regOpenKeyEx(hKey: HKEY, lpSubKey: WideCString, ulOptions: int32,
+                  samDesired: int32,
+                  phkResult: var HKEY): int32 {.
+  importc: "RegOpenKeyExW", dynlib: "Advapi32.dll", stdcall.}
+
+proc regCloseKey(hkey: HKEY): int32 {.
+  importc: "RegCloseKey", dynlib: "Advapi32.dll", stdcall.}
+
+proc regGetValue(key: HKEY, lpSubKey, lpValue: WideCString;
+                 dwFlags: int32 = RRF_RT_ANY, pdwType: ptr int32,
+                 pvData: pointer,
+                 pcbData: ptr int32): int32 {.
+  importc: "RegGetValueW", dynlib: "Advapi32.dll", stdcall.}
+
+proc regQueryInfoKey( hKey: HKEY,
+                       lpClass : WideCstring,
+                       lpcchClass : ptr int32,
+                       lpReserved : ptr int32,
+                       lpcSubKeys : ptr int32,
+                       lpcbMaxSubKeyLen : ptr int32,
+                       lpcbMaxClassLen : ptr int32,
+                       lpcValues : ptr int32,
+                       lpcbMaxValueNameLen : ptr int32,
+                       lpcbMaxValueLen : ptr int32,
+                       lpcbSecurityDescriptor : ptr int32,
+                       lpftLastWriteTime : ptr FILETIME): int32 {.
+  importc: "RegQueryInfoKeyW", dynlib: "Advapi32.dll", stdcall.}
+
+proc regEnumValue( hKey: HKEY,
+                    dwIndex : int32,
+                    lpValueName : WideCString,
+                    lpcchValueName : ptr int32,
+                    lpReserved : ptr int32,
+                    lpType : ptr int32,
+                    lpData : pointer,
+                    lpcbData : ptr int32): int32 {.
+  importc: "RegEnumValueW", dynlib: "Advapi32.dll", stdcall.}
+
+template call(f) =
+  let err = f
+  if err != 0:
+    raiseOSError(err.OSErrorCode, astToStr(f))
+
+iterator enumKeyValues*(path: string, handle: HKEY): tuple[key:string, value:string] =
```

Comment:
Might be a good idea if we can get the Nim standard library to include something like this in the future, I can see it being quite a useful iterator!

Reply:
Wow high praise indeed!  I've asked the question on the forum: [https://forum.nim-lang.org/t/5678#35245](https://forum.nim-lang.org/t/5678#35245)